add codegen subdirectories to mypy-strict.ini

bdhirsh opened the pull request:
I might just not be using the correct workflow, but when I make changes locally I check mypy by running `mypy --config=mypy-strict.ini` (I got that command from looking at what CI does).

It looks like that config file doesn't recursively check all of the python files in the codegen folder, so I added the subfolders directly into the config: `tools/codegen/api` and `tools/codegen/dest`.

What's a little weird to me is the CI mypy tests would catch mypy errors in those files even though my local command would pass, so maybe CI is doing something a little different.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56523 add codegen subdirectories to mypy-strict.ini**

Differential Revision: [D27890855](https://our.internmc.facebook.com/intern/diff/D27890855)

bdhirsh committed:
add codegen subdirectories to mypy-strict.ini

[ghstack-poisoned]

samestep commented on mypy-strict.ini:
```diff
@@ -40,6 +40,8 @@ files =
     tools/autograd/*.py,
     tools/clang_tidy.py,
     tools/codegen/*.py,
+    tools/codegen/api/*.py,
+    tools/codegen/dest/*.py,
```
Rather than using a wildcard that only matches one directory deep, why not just match the entire directory?
```suggestion
    tools/codegen/api,
    tools/codegen/dest,
```
See #55700 which does a similar thing.

samestep replied:
Actually, you might want to ignore this and instead use my other suggestion.

samestep reviewed (commented):
Looks good, thanks! I left one suggestion just to make sure I understand what your goal is here.

> I might just not be using the correct workflow, but when I make changes locally I check mypy by running `mypy --config=mypy-strict.ini` (I got that command from looking at what CI does).

That sounds right; I'm assuming your `mypy` version is correct? (You should see a warning message if it isn't.)

> It looks like that config file doesn't recursively check all of the python files in the codegen folder, so I added the subfolders directly into the config: tools/codegen/api and tools/codegen/dest.
> 
> What's a little weird to me is the CI mypy tests would catch mypy errors in those files even though my local command would pass, so maybe CI is doing something a little different.

That is strange; it should be the same in CI as local. I saw the same behavior while working on #55700: if you replace `benchmarks/instruction_counts` with just `benchmarks/instruction_counts/*.py` (which only matches `benchmarks/instruction_counts/main.py` and nothing in the deeper subdirectories), it will still typecheck the Python files in those deeper subdirectories because that `main.py` file imports them (some transitively).

samestep commented on mypy-strict.ini:
```diff
@@ -40,6 +40,8 @@ files =
     tools/autograd/*.py,
     tools/clang_tidy.py,
     tools/codegen/*.py,
+    tools/codegen/api/*.py,
+    tools/codegen/dest/*.py,
```
Actually, I now see that you could just do this instead, which is even simpler:
```suggestion
    tools/codegen,
```

bdhirsh replied:
ah yep, way better thanks!

bdhirsh committed:
Update mypy-strict.ini

Co-authored-by: Sam Estep <sestep@fb.com>

samestep commented:
As an aside, I'm starting to think that _all_ the `foo/bar/*.py` patterns in `mypy.ini` and `mypy-strict.ini` might be a mistake and actually intend to mean just `foo/bar`; @pytorch/pytorch-dev-infra thoughts?

samestep commented:
See also #55702 for more context.

bdhirsh commented:
> I saw the same behavior while working on #55700: if you replace `benchmarks/instruction_counts` with just `benchmarks/instruction_counts/*.py` (which only matches `benchmarks/instruction_counts/main.py` and nothing in the deeper subdirectories), it will still typecheck the Python files in those deeper subdirectories because that `main.py` file imports them (some transitively).

Yeah, that's probably what's happening here then on CI, since `tools/codegen/gen.py` imports most/all of the files in those subfolders. It sounds like I should have expected to see that behavior locally too, but either way this PR should fix it :) thanks!

bdhirsh committed:
Update on "add codegen subdirectories to mypy-strict.ini"

I might just not be using the correct workflow, but when I make changes locally I check mypy by running `mypy --config=mypy-strict.ini` (I got that command from looking at what CI does).

It looks like that config file doesn't recursively check all of the python files in the codegen folder, so I added the subfolders directly into the config: `tools/codegen/api` and `tools/codegen/dest`.

What's a little weird to me is the CI mypy tests would catch mypy errors in those files even though my local command would pass, so maybe CI is doing something a little different.


Differential Revision: [D27890855](https://our.internmc.facebook.com/intern/diff/D27890855)

[ghstack-poisoned]
